nodejs/stream_writer.ts

```diff
@@ -0,0 +1,81 @@
+// Copyright Pravega Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Native modules are not currently supported with ES module imports.
+// https://nodejs.org/api/esm.html#esm_no_native_module_loading
+import { createRequire } from 'module';
+const require = createRequire(import.meta.url);
+
+const { StreamWriterWriteEventBytes, StreamWriterFlush, StreamWriterToString } = require('./index.node');
+
+/**
+ * A writer for a stream.
+ * Note: A StreamWriter cannot be created directly without using the StreamManager.
+ */
+export interface StreamWriter {
+    /**
+     * Write an event into the Pravega Stream. The events that are written will appear
+     * in the Stream exactly once. The event of type string is converted into bytes with `UTF-8` encoding.
+     *
+     * Note that the implementation provides retry logic to handle connection failures and service host
+     * failures. Internal retries will not violate the exactly once semantic so it is better to rely on them
+     * than to wrap this with custom retry logic.
+     *
+     * @param event String that want to be written to the Pravega Stream.
+     * @param routing_key (optional) The user specified routing key.
+     * @returns Promise<void>
+     */
+    write_event: (event: string, routing_key?: string) => Promise<void>;
+
+    /**
+     * Write a byte array into the Pravega Stream. This is similar to `write_event(...)` api except
+     * that the the event to be written is a byte array.
+     *
+     * Note that the implementation provides retry logic to handle connection failures and service host
+     * failures. Internal retries will not violate the exactly once semantic so it is better to rely on them
+     * than to wrap this with custom retry logic.
+     *
+     * @param buf Byte array that want to be written to the Pravega Stream.
+     * @param routing_key (optional) The user specified routing key.
+     * @returns Promise<void>
+     */
+    write_event_bytes: (buf: Uint8Array, routing_key?: string) => Promise<void>;
```

Comment:
How does this work when reading? Does the reader have to know which way the event was written or is it always read back as bytes? If the reader has to know, then it might be better to have two types of writer so that one isn't tempted to mix them together on the same stream which would be impossible for a reader to deal with.

Reply:
There is only one `js_write_event_bytes` on the rust side, so everything is written in bytes. `write_event` in js/ts is only a helper method that calls a js/ts `TextEncoder` to convert string to bytes in utf-8. For the reader side, users need to use the `TextDecoder('utf-8')` in js/ts to deserialize the bytes to string.